sycl/doc/extensions/proposed/sycl_ext_oneapi_peer_access.asciidoc

```diff
@@ -0,0 +1,162 @@
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+
+=== Peer to Peer (P2P) Memory Access APIs
+
+This extension adds support for mechanisms to query and enable support for
+direct memory access between peer devices in a system.
+In particular, this allows one device to directly access USM Device
+allocations for a peer device in the same context.
+Peer to peer capabilities are useful as they can provide access to a peer
+device's memory inside a compute kernel and also optimized memory copies between
+peer devices.
+
+This extension adds the following new member functions to the device class, as described
+below.
+
+[source,c++]
+----
+namespace sycl   {
+namespace ext    {
+namespace oneapi {
+enum class peer_access {
+  access_supported,
+  access_enabled,
+  atomics_supported,
+};
+} // namespace oneapi
+} // namespace ext
+
+class device {
+public:
+  bool ext_oneapi_can_access_peer(const device &peer,
+                                  ext::oneapi::peer_access value =
+                                    ext::oneapi::peer_access::access_supported);
+  void ext_oneapi_enable_peer_access(const device &peer);
+  void ext_oneapi_disable_peer_access(const device &peer);
+};
+
+} // namespace sycl
+----
+
+The semantics of the new functions are:
+
+|===
+|Member Function |Description
+
+|bool ext_oneapi_can_access_peer(const device &peer,
+                                  ext::oneapi::peer_access value =
+                                  ext::oneapi::peer_access::access_supported)
+a|Queries the peer access status between this device and `peer` according to
+the query `value`:
+
+* `ext::oneapi::peer_access::access_supported`: Returns true only if it is
+possible for this device to enable peer access to USM device memory allocations
+located on the `peer` device.
+
+* `ext::oneapi::peer_access::access_enabled`: Returns true only if peer access is
+currently enabled from this device to the `peer` device.
+
+* `ext::oneapi::peer_access::atomics_supported`: When this query returns true,
+it indicates that this device may concurrently perform atomic operations on USM
+device memory allocations located on the `peer` device when peer access is enabled
+to that device. Atomics performed on a peer device's memory must have
+`memory_scope::system` scope.
+If the query returns false, attempting to perform atomic operations
+on `peer` memory will have undefined behavior.
+
+|void enable_peer_access(const device &peer)
+|Enables this device to access USM device allocations located on the peer
+device. This does not permit the peer device to access this device's memory.
+Once this access is enabled, SYCL kernel functions and the explicit memory
+functions may access USM device allocations on the peer device subject to the
+normal rules about context as described in the core SYCL specification.
+Throws an exception if access cannot be enabled or if access is already
+enabled.
```

Comment:
This should clarify which exception is thrown.  Something like:

> If this device does not support peer access (as defined by `peer_access::access_supported`), throws an `exceptions` with the `errc::feature_not_supported` error code.  If access is already enabled, throws an `exception` with the `errc::invalid` error code.

Can you refresh my memory, though, about why we want to throw an exception if access is already supported?  Couldn't the function just do nothing in that case?

Reply:
Currently this happens in this case in the draft impl:

```
PI CUDA ERROR:
	Value:           704
	Name:            CUDA_ERROR_PEER_ACCESS_ALREADY_ENABLED
	Description:     peer access is already enabled
	Function:        cuda_piextEnablePeer
	Source Location: llvm/sycl/plugins/cuda/pi_cuda.cpp:5625
```